File: impl/rb/lib/predicator.rb
require "predicator/context"
require "predicator/evaluator"
require "predicator/parser"

module Predicator
  def self.parse source
    Predicator::Parser.new.parse source
  end

  def self.compile source
    ast = parse source
    ast.to_instructions
  end

  def self.evaluate source, context={}
    instructions = compile source
    evaluate_instructions instructions, context
  end

  def self.evaluate_instructions instructions, context={}
    evaluator = Evaluator.new instructions, context
    evaluator.result
  end
end

File: impl/ts/src/predicator.js
// ----------------- Lexer -----------------
const { tokenize } = require('./lexer')
const {
  PredicatorParser,
  parserInstance,
  parse,
  BasePredicatorCstVisitor,
  BasePredicatorCstVisitorWithDefaults
} = require('./parser')
const {
  PredicatorEvaluator,
  evaluate,
  evaluateInstructions
} = require('./evaluator')

const { toInstructions } = require('./visitors/instructions')

module.exports = {
  tokenize,

  PredicatorParser,
  parserInstance,
  parse,
  BasePredicatorCstVisitor,
  BasePredicatorCstVisitorWithDefaults,

  PredicatorEvaluator,
  evaluate,
  evaluateInstructions,

  toInstructions
}
